constraint/pkg/client/drivers/local/local.go

```diff
@@ -75,22 +85,82 @@ func New(args ...Arg) drivers.Driver {
 	for _, arg := range args {
 		arg(d)
 	}
+
+	// adding externaldata builtin otherwise capabilities get overridden
+	// if a capability, like http.send, is disabled
+	if d.providerCache != nil {
+		d.capabilities.Builtins = append(d.capabilities.Builtins, &ast.Builtin{
+			Name: "externaldata",
+			Decl: opatypes.NewFunction(opatypes.Args(opatypes.S, opatypes.S), opatypes.A),
+		})
+	}
 	d.compiler.WithCapabilities(d.capabilities)
 	return d
 }
 
 var _ drivers.Driver = &driver{}
 
 type driver struct {
-	modulesMux   sync.RWMutex
-	compiler     *ast.Compiler
-	modules      map[string]*ast.Module
-	storage      storage.Store
-	capabilities *ast.Capabilities
-	traceEnabled bool
+	modulesMux    sync.RWMutex
+	compiler      *ast.Compiler
+	modules       map[string]*ast.Module
+	storage       storage.Store
+	capabilities  *ast.Capabilities
+	traceEnabled  bool
+	providerCache *externaldata.ProviderCache
 }
 
 func (d *driver) Init(ctx context.Context) error {
+	if d.providerCache != nil {
+		rego.RegisterBuiltin2(
+			&rego.Function{
+				Name:    "externaldata",
+				Decl:    opatypes.NewFunction(opatypes.Args(opatypes.S, opatypes.S), opatypes.A),
+				Memoize: true,
+			},
+			func(bctx rego.BuiltinContext, a, b *ast.Term) (*ast.Term, error) {
+				var providerName, body string
+
+				if err := ast.As(a.Value, &providerName); err != nil {
+					return nil, err
+				}
+				if err := ast.As(b.Value, &body); err != nil {
+					return nil, err
+				}
+
+				provider, err := d.providerCache.Get(providerName)
+				if err != nil {
+					return nil, fmt.Errorf("unable to retrieve provider %v cache", providerName)
+				}
+
+				req, err := http.NewRequest("GET", provider.Spec.ProxyURL, bytes.NewBuffer([]byte(body)))
```

Comment:
It looks like we're just passing the value through wholesale. Do we want to define a protocol here so we can preserve the ability to pass things like additional data/context or do batching?

Reply:
Are you thinking additional data and batching could be part of the provider spec that can be evaluated here before making the http request?

Reply:
I'm thinking it's an important part of the requester/provider protocol we're establishing that we are missing since we aren't defining anything about what a request/response between Gatekeeper and the provider would look like.

A (kind of) related example: consider that K8s admission requests have structured request/response objects that allow us to write a webhook that is compatible across K8s versions and allows them to do things like add additional request data without breaking us and also allows them to interpret our results in a consistent manner. Since we are also creating a REST API, we should strive for the same kinds of backwards compatibility.

Two themes I'm thinking of here:

- We should define a stable protocol so that people can write providers and we can support them in a backward-compatible manner even if we introduce new features, such as (not saying all of these will necessarily happen, but they are common REST extensions):
  - Pagination
  - Batching
  - Auth (assuming bearer token)
  - Cache/resumption keys
  - Error reasons/ failure-mode hints (e.g. it's worth retrying, downstream problem, etc.), similar to how K8s added a "warning" severity to admission webhook responses.
  - Permissions delegation tokens
  - Common request context (e.g. requesting user)
- Having a well-defined place for parameters to be passed allows us to cache based off of the `(provider, parameter)` tuple. Caching is only possible if we know an answer will be consistent for a given request and we only know that if we can know where all the variables that could change a response are provided. Examples:
  - assuming we have configurable timeouts in the future, it should be possible to send a request with the same parameters but a different timeout and still get a cache hit.
  - In the case of different batches of requests, we should be able to unpack the keys being requested and see if they've already been cached. If I have requested `[1, 2, 3, 4]` before, and now I'm requesting `[1, 2]`, we should be able to serve that from the cache.

There may be more, but those are top-of-mind for me.

One example in the corresponding G8r PR:

```go
func SendProviderRequest(provider *externaldatav1alpha1.Provider, providerResponseCache map[types.ProviderCacheKey]string) (map[types.ProviderCacheKey]string, error) {

	body, err := json.Marshal(providerResponseCache)
 
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("GET", provider.Spec.ProxyURL, bytes.NewBuffer(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	timeout := time.Second * time.Duration(provider.Spec.Timeout)
	client := &http.Client{Timeout: timeout}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var respBody []byte
	if resp.StatusCode == 200 {
		respBody, err = ioutil.ReadAll(resp.Body)
		if err != nil {
			log.Error(err, "unable to read response body")
			return nil, err
		}
	}

	var result map[types.ProviderCacheKey]string
	err = json.Unmarshal(respBody, &result)
	if err != nil {
		return nil, err
	}

	return result, nil
}
```

We are exporting our cache directly to the provider, which implies that a provider should accept a format of `map[ProviderCacheKey]string` where `ProviderCacheKey` is defined as:

```go
// ProviderCacheKey is the map key for provider requests.
type ProviderCacheKey struct {
	ProviderName string `json:"providerName,omitempty"`
	// outbound data is based on Provider DataSource
	// it is either value of location or username in admission request
	OutboundData string     `json:"outboundData,omitempty"`
	DataSource   DataSource `json:"dataSource,omitempty"`
}
```

This means that:

- Rego code that wants to call the same provider must construct this same `map[ProviderCacheKey]string` request format (or else the Rego function needs to construct it), which pushes complexity on to Rego authors and could be problematic as `DataSource` only has meaning for mutation.
- We are coupling provider implementation (handled by an outside party) to our internal cache format. This means:
  - If we ever change the format of our cache, we risk breaking pre-existing providers
  - If we cannot break pre-existing providers, we are limited in how we change our cache format, which may limit our ability to improve performance or support different kinds of providers in the future
  - The request/response size grows with the size of the cache, which is bandwidth/latency/CPU inefficient (request size, deserialization cost), and leads to an overall N^2 complexity for total data sent per evaluation, where N is the number of mutators. A protocol that avoids exporting the cache would avoid this.
- As-written, we are creating a de-facto protocol that provider implementors (and Rego authors) need to be aware of, but it's not documented. By defining an explicit request/response interface, we can provide clarity and stability, and insulate Rego authors from the caching complexity.

Reply:
added an API for ProviderRequest and ProviderResponse which is decoupled from the rego request and response types.